### PR TITLE
Subscriptions + Elements for Stripe payments

### PR DESCRIPTION
**Pull request by chdecultot**
Hi,

This PR is a proposal to enhance the current Stripe payment gateway in two ways:

1. Add the possibility to send payment requests for subscriptions
A new field in the payment request allows a user to select a payment plan:
image
Frappe PR: https://github.com/frappe/frappe/pull/5602

I have created a payment plan doctype for this purpose. The plan is keep a generic payment plan doctype that could be used by other payment platforms like Braintree or Razorpay.
image

2. Move from Stripe Checkout to Stripe Elements (much more configurable and elegant):
image

Thanks in advance for your feedback!

### CODE REVIEW

File: erpnext/docs/user/manual/en/setting-up/integrations/stripe-integration.md [hunk 1]
```diff
@@ -18,6 +18,21 @@ Also it will create Payment Gateway Account entry. Payment Gateway Account is co
 
 After configuring Payment Gateway Account your system is able to accept online payments.
 
+### Setup subscriptions plans
+
+If you need to bill a recurring amount instead of a one-time charge, you can use Stripe's subscription system.
```
Comment (saurabh6790) [inline erpnext/docs/user/manual/en/setting-up/integrations/stripe-integration.md]: We have shifted Docs to Foundation app. Please send a PR to foundation app. https://github.com/erpnext/foundation

File: erpnext/erpnext_integrations/doctype/payment_plan/payment_plan.json [hunk 2]
```diff
@@ -0,0 +1,195 @@
+{
```
Comment (saurabh6790) [inline erpnext/erpnext_integrations/doctype/payment_plan/payment_plan.json]: Payment plan and Subscription Plans are same. Please add Payment Plan Id to subscription Id. Also add a description for field payment_plan_id.

File: erpnext/erpnext_integrations/doctype/payment_plan/payment_plan.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import frappe
+from frappe import _
+from frappe.integrations.utils import create_request_log
+import stripe
+
+class PaymentPlan(Document):
+	pass
+
+
+def create_stripe_subscription(gateway_controller, data):
```
Comment (saurabh6790) [inline erpnext/erpnext_integrations/doctype/payment_plan/payment_plan.py]: Under erpnext_integration, create a new file stripe.py to manage erpnext sides actions.

File: erpnext/erpnext_integrations/doctype/stripe_settings/stripe_settings.js [hunk 4]
```diff
@@ -0,0 +1,6 @@
+// Copyright (c) 2018, Frappe Technologies Pvt. Ltd. and contributors
+// For license information, please see license.txt
+
+frappe.ui.form.on('Stripe Settings', {
```
Comment (saurabh6790) [inline erpnext/erpnext_integrations/doctype/stripe_settings/stripe_settings.js]: Already exists in frappe

File: erpnext/erpnext_integrations/doctype/stripe_settings/stripe_settings.json [hunk 5]
```diff
@@ -0,0 +1,315 @@
+{
```
Comment (saurabh6790) [inline erpnext/erpnext_integrations/doctype/stripe_settings/stripe_settings.json]: Already exists in frappe!

File: erpnext/erpnext_integrations/doctype/stripe_settings/stripe_settings.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
+# -*- coding: utf-8 -*-
```
Comment (saurabh6790) [inline erpnext/erpnext_integrations/doctype/stripe_settings/stripe_settings.py]: Already exists in frappe. If any particular code related to ERPNext, create a new file stripe.py under erpnext_integrations and add the ERPNext specific code.

File: erpnext/erpnext_integrations/doctype/stripe_settings/stripe_settings.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+		create_payment_gateway('Stripe-' + self.gateway_name, settings='Stripe Settings', controller=self.gateway_name)
+		call_hook_method('payment_gateway_enabled', gateway='Stripe-' + self.gateway_name)
+		if not self.flags.ignore_mandatory:
+			self.validate_stripe_credentails()
+
+	def validate_stripe_credentails(self):
+		if self.publishable_key and self.secret_key:
+			header = {"Authorization": "Bearer {0}".format(self.get_password(fieldname="secret_key", raise_exception=False))}
+			try:
+				make_get_request(url="https://api.stripe.com/v1/charges", headers=header)
```
Comment (saurabh6790) [inline erpnext/erpnext_integrations/doctype/stripe_settings/stripe_settings.py]: If using strip's python SDK, also handle validations with SDK. Else Use REST API'S in all scope.

**Review by saurabh6790 — CHANGES_REQUESTED** [review]

**Comment by saurabh6790** [comment]
@chdecultot Thank you for the PR. Please go through the suggestions.

**Comment by nabinhait** [comment]
@chdecultot Closing the for now, please reopen after fixing.

**Pull request closed without merging**